File: ocaml/lib/ema.ml
module EMA = struct
  type t = {
    mutable b: float;
    mutable n: float;
    mutable avg: float;
  }

  let create ?(b = 0.9) () = {
    b;
    n = 0.;
    avg = 0.;
  }

  let next ema v =
    ema.n <- ema.n +. 1.;
    ema.avg <- ema.b *. v +. (1. -. ema.b) *. ema.avg;
    ema.avg

  let avg ema = ema.avg
end

File: go/ema.go
package futures

type EMA struct {
    B   float64
    N   float64
    Avg float64
}

func NewEMA() *EMA {
    return &EMA{
        B:   0.9,
        N:   0,
        Avg: 0,
    }
}

func (e *EMA) Next(v float64) float64 {
    e.N++
    e.Avg = e.B*v + (1-e.B)*e.Avg
    return e.Avg
}

func (e *EMA) Avg() float64 {
    return e.Avg
}